File: BaekJoon/Gold/11404/Python/11404.py
# no. 11404: 플로이드 (Gold IV)
# 모든 도시의 쌍 (A, B)에 대해서 도시 A에서 B로 가는데 필요한 비용의 최소값
# 플로이드 와샬 알고리즘
# dist[n][m] = min(dist[n][m], dist[n][k]+dist[k][m]) 이라는 논리
# 이때 최단거리를 구해야 하므로 k가 1부터 n까지의 모든 도시인 경우를 탐색해야한다.
# 따라서 시간복잡도는 O(N^3)

import sys
from math import inf
input = sys.stdin.readline

n = int(input())
m = int(input())
dist = [[inf for _ in range(n)] for _ in range(n)]

for _ in range(m):
    start, end, cost = map(int, input().split())
    start -= 1
    end -= 1
    dist[start][end] = min(dist[start][end], cost)

for i in range(n):
    dist[i][i] = 0

for k in range(n):
    for start in range(n):
        for end in range(n):
            dist[start][end] = min(dist[start][end], dist[start][k]+dist[k][end])

for i in range(n):
    for j in range(n):
        print(dist[i][j] if dist[i][j] != inf else 0, end=' ')
    print()

File: BaekJoon/Gold/11404/C++/11404.cpp
// no.11404: 플로이드 (G4)

#include <cstdio>
#include <vector>
using namespace std;

int main() {
    int n, m;
    scanf("%d", &n);
    scanf("%d", &m);
    vector<vector<int>> dist(n+1, vector<int>(n+1, -1));
    for(int i=1; i<=n; i++) {
        dist[i][i] = 0;
    }
    for(int i=0; i<m; i++) {
        int a, b, c;
        scanf("%d %d %d", &a, &b, &c);
        if(dist[a][b] == -1 || c < dist[a][b])
            dist[a][b] = c;
    }
    for(int k=1; k<=n; k++) {
        for(int i=1; i<=n; i++) {
            for(int j=1; j<=n; j++) {
                if(dist[i][k] == -1 || dist[k][j] == -1)
                    continue;
                if(dist[i][j] == -1 || dist[i][k] + dist[k][j] < dist[i][j])
                    dist[i][j] = dist[i][k] + dist[k][j];
            }
        }
    }
    for(int i=1; i<=n; i++) {
        for(int j=1; j<=n; j++) {
            printf("%d ", dist[i][j] == -1 ? 0 : dist[i][j]);
        }
        printf("\n");
    }
    return 0;
}